File: 1428-jump-game-iii/jump-game-iii.go
func canReach(arr []int, start int) bool {
    visited := make([]bool, len(arr))
	return dp(arr, start, visited)
}

func dp(arr []int, idx int, visited []bool) bool {
	if idx < 0 || idx >= len(arr) {
		return false
	}

	if visited[idx] {
		return false
	}

	if arr[idx] == 0 {
		return true
	}

	visited[idx] = true

	left := dp(arr, idx-arr[idx], visited)
	right := dp(arr, idx+arr[idx], visited)

	return left || right
}

File: 1428-jump-game-iii/jump-game-iii.java
class Solution {
    int[] arr;
    HashSet<Integer> set = new HashSet<>();
    public boolean canReach(int[] arr, int start) {
        this.arr = arr;
       return dfs(start);
    }

    public boolean dfs(int idx){
        if(this.set.contains(idx)){
            return false;
        }

        this.set.add(idx);

        if(idx < 0 || idx >= this.arr.length){
            return false;
        }
        if(this.arr[idx] == 0){
            return true;
        }

        boolean left = dfs(idx - this.arr[idx]);
        boolean right = dfs(idx + this.arr[idx]);

        return left || right;
    }
}